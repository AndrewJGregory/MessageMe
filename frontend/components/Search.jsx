import React from "react";
import UserSearchResult from "./UserSearchResult";
import { withRouter } from "react-router-dom";

class Search extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      query: "",
      textAlignment: "",
      shouldUsersBeDisplayed: false
    };
    this.updateInput = this.updateInput.bind(this);
    this.leftAlignText = this.leftAlignText.bind(this);
    this.handleOnBlur = this.handleOnBlur.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }

  updateInput(e) {
    e.preventDefault();
    const query = e.target.value;
    this.setState({ query });
  }

  componentDidMount() {
    this.props.fetchUsers("");
    this.props.createChat(
      this.props.currentUserId,
      this.props.match.params.userId
    );
  }

  leftAlignText(e) {
    e.preventDefault();
    this.setState({
      textAlignment: "left-text-align",
      shouldUsersBeDisplayed: true
    });
  }

  handleOnBlur(e) {
    e.preventDefault();
    this.setState({ textAlignment: "", shouldUsersBeDisplayed: false });
  }

  handleClick(e) {
    e.preventDefault();
    const id = e.target.dataset.userId;
    this.props.history.push(`/messages/${id}`);
    this.props.createChat(this.props.currentUserId, id);
  }

  render() {
    const regExp = new RegExp(`${this.state.query}`);
    let users = null;
    users = this.props.userResults.reduce((users, user) => {
      if (regExp.test(user.username)) {
        users.push(<UserSearchResult user={user} key={user.id} />);
      }
      return users;
    }, []);

    return (
      <div className="search">
        <div className="user-search-input-container">
          <i className="fa fa-search" />
          <input
            type="text"
            onChange={this.updateInput}
            value={this.state.query}
            placeholder="Search MessageMe"
            className="user-search-input"
            id={`${this.state.textAlignment}`}
            onFocus={this.leftAlignText}
            onBlur={this.handleOnBlur}
          />
        </div>
        <ul className="user-search-results" onClick={this.handleClick}>
          {users}
        </ul>
      </div>
    );
  }
}

export default withRouter(Search);
