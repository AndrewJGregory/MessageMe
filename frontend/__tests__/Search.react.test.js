import React from "react";
import Search from "../components/Search";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer
    .create(<Search userResults={[]} fetchUsers={jest.fn} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
