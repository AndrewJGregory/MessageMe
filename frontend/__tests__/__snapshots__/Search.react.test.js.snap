// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="search"
>
  <div
    className="user-search-input-container"
  >
    <i
      className="fa fa-search"
    />
    <input
      className="user-search-input"
      id=""
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Search MessageMe"
      type="text"
      value=""
    />
  </div>
  <ul
    className="user-search-results"
  />
</div>
`;
